### PR TITLE
Add a Godot 3 to Godot 4 upgrade guide

### PR DESCRIPTION
**Pull request by Calinou**
I've likely missed some things, so please review :slightly_smiling_face:

See https://github.com/godotengine/godot-docs/issues/5121. This closes https://github.com/godotengine/godot-docs/issues/4960.

### CODE REVIEW

**Comment by Mickeon** [comment]
A lot of properties, methods, signals etc. etc. are missing and I'm not sure it's because they are handled by the converter already or is an intentional omission to keep it brief.

**Comment by Calinou** [comment]
> A lot of properties, methods, signals etc. etc. are missing and I'm not sure it's because they are handled by the converter already or is an intentional omission to keep it brief.

In the interest of brevity, I've only listed properties/methods that are *not* automatically converted. I've left a link to the converter source code, which lists all properties/methods that were renamed (including commented out code for properties/methods that are not automatically renamed, with a comment explaining why).

**Comment by Mickeon** [comment]
I was considering reordering that whole list 'cause it very much is... such a mess to read, but in the meantime, let's see...

**Comment by Mickeon** [comment]
- Camera2D's `rotating` has been reverted the other way around https://github.com/godotengine/godot/pull/64359;
- Button's signals are now `button_up` and `button_down` and the property is `button_pressed` (_wait what, why is it not just `pressed` anymore after the whole ordeal?_)
- https://github.com/godotengine/godot/pull/66017
- https://github.com/godotengine/godot/pull/65188 (_I recall Reduz wanted to nuke this class, but well, it is still here, so it may be worth mentioning_)
- https://github.com/godotengine/godot/pull/65038 and https://github.com/godotengine/godot/pull/64665 are two entirely different properties with the same name, and for that reason could not be included in the project converter.
- https://github.com/godotengine/godot/pull/64661 (_This is worth mentioning because the converter cannot infer if it refers to TileMap or GridMap_)
- https://github.com/godotengine/godot/pull/64377 (_update is too generic for a term and could not be included in the project converter_)
- https://github.com/godotengine/godot/pull/65148 may be worth mentioning? The original PR's behavior was modified so it isn't nearly as compatibility breaking as thought, unless someone was **actually** relying on the value being clamped to 0.

**Comment by Calinou** [comment]
@Mickeon Thanks! I've pushed a separate commit mentioning your changes so it can be reviewed separately.

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
++---------------------+-----------------------+----------------------------------------------------------------------+
+| BitmapFont          | FontFile              | See :ref:`doc_gui_using_fonts`.                                      |
++---------------------+-----------------------+                                                                      |
+| DynamicFont         | FontFile              |                                                                      |
++---------------------+-----------------------+                                                                      |
+| DynamicFontData     | FontFile              |                                                                      |
++---------------------+-----------------------+----------------------------------------------------------------------+
+| Navigation3D        | Node3D                | Replaced by other Navigation nodes.                                  |
++---------------------+-----------------------+----------------------------------------------------------------------+
+| Navigation2D        | Node2D                | Replaced by other Navigation nodes.                                  |
```
Comment (skyace65) [inline tutorials/editor/upgrading_to_godot_4.rst:469]: I'd add a link to the [2D ](https://docs.godotengine.org/en/latest/tutorials/navigation/navigation_introduction_2d.html)and [3D ](https://docs.godotengine.org/en/latest/tutorials/navigation/navigation_introduction_3d.html)navigation overview pages here.

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+hundreds in large projects). This is because the project upgrade tool cannot
+cater to all situations. Therefore, a large part of the upgrade process remains
+manual.
+
+Automatically renamed nodes and resources
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The list below refers to the nodes' new names in Godot 4.0. In the transition
+between Godot 3.x and 4.0, dozens of nodes were renamed. 3D nodes had a ``3D``
+prefix added to them for consistency with their 2D counterparts. This conversion
```
Comment (skyace65) [inline tutorials/editor/upgrading_to_godot_4.rst:100]:
```suggestion
suffix added to them for consistency with their 2D counterparts. This conversion
```

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
++-----------------------------------------+-------------------------------------------+
+| Physics2DServer                         | PhysicsServer2D                           |
++-----------------------------------------+-------------------------------------------+
+| Physics2DShapeQueryParameters           | PhysicsShapeQueryParameters2D             |
++-----------------------------------------+-------------------------------------------+
+| Physics2DTestMotionResult               | PhysicsTestMotionResult2D                 |
++-----------------------------------------+-------------------------------------------+
+| PhysicsDirectBodyState                  | PhysicsDirectBodyState3D                  |
++-----------------------------------------+-------------------------------------------+
+| PhysicsDirectSpaceState                 | PhysicsDirectSpaceState3D                 |
```
Comment (skyace65) [inline tutorials/editor/upgrading_to_godot_4.rst:173]: This and DirectBodyState were only changed with a suffix being added and should be removed

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+Updating scripts to take backwards-incompatible changes into account
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some changes performed between Godot 3.x and 4 are not renames, but they still
+break backwards compatibility due to different default behavior.
+
+The most notable examples of this are:
+
+- Built-in scripts that are :ref:`tool scripts <doc_running_code_in_the_editor>`
+  do not get the the ``tool`` keyword converted to the ``@tool`` annotation.
```
Comment (skyace65) [inline tutorials/editor/upgrading_to_godot_4.rst:421]:
```suggestion
  do not get the ``tool`` keyword converted to the ``@tool`` annotation.
```

**Review by skyace65 — CHANGES_REQUESTED** [review]
I accidentally hit submit review with the first thing I was looking at, sorry. Just a few minor things that should be changed but overall the page the page looks good. As someone not familiar with a lot of the changes made this was easy to understand.

**Comment by Mickeon** [comment]
Some miscellaneous compatibility-breaking PRs I found while browsing.
- https://github.com/godotengine/godot/pull/66133
- https://github.com/godotengine/godot/pull/65437
- https://github.com/godotengine/godot/pull/64952 (This may be noteworthy on its own?)
- https://github.com/godotengine/godot/pull/64412 (?)
- https://github.com/godotengine/godot/pull/64417
- https://github.com/godotengine/godot/pull/61647
- https://github.com/godotengine/godot/pull/63595 (Very very important)
- https://github.com/godotengine/godot/pull/59589 (There's also noslider to no_slider but I can't find it)
- https://github.com/godotengine/godot/pull/51591

**Review by skyace65 — APPROVED** [review]

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    conversion tool found in the Godot 4 editor.
+
+    Projects made using Godot versions older than 3.0 must follow a 2-step
+    process:
+
+    1. Upgrade the project to Godot 3 using the **Tools > Export to Godot 3.0**
+       option in the Godot 2.1.6 editor. Open the upgraded project in the Godot
+       3.x editor, then manually fix the project to make sure it looks and
+       runs correctly.
+    2. Only *after* performing step 1, you may upgrade the project to Godot 4.
```
Comment (akien-mga) [inline tutorials/editor/upgrading_to_godot_4.rst:84]: I'm not sure this is worth describing, it literally affects maybe 2 users, who know well what they would have to do to convert to 3.x if wanted.

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+5. Wait until the project conversion process finishes. This can take up to a few
+   minutes for large projects with lots of scenes.
+6. When the project manager interface becomes available again, double-click the
+   project (or select the project then choose **Edit**) to open it in the
+   editor.
+
+.. note::
+
+    Only Godot 3.0 and later projects can be upgraded using the project
+    conversion tool found in the Godot 4 editor.
```
Comment (akien-mga) [inline tutorials/editor/upgrading_to_godot_4.rst:75]:
```suggestion
    Only Godot 3.0 and later projects can be upgraded using the project
    conversion tool found in the Godot 4 editor.
    It is recommended to ensure that your project is up-to-date with the
    latest 3.x stable release before running the project upgrade tool.
```

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 7]
```diff
@@ -0,0 +1,540 @@
+.. _doc_upgrading_to_godot_4:
+
+Upgrading to Godot 4
+====================
+
+TODO: https://github.com/godotengine/godot-proposals/issues/387#issuecomment-1251909261
```
Comment (akien-mga) [inline tutorials/editor/upgrading_to_godot_4.rst:6]: It's a bit weird as the first thing in the document.

If this is meant to link to further information that can be useful when doing manual porting work, I would link it as a `seealso` further down, where the user would typically need to dig deeper.

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+Automatically renamed nodes and resources
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The list below refers to the nodes' new names in Godot 4.0. In the transition
+between Godot 3.x and 4.0, dozens of nodes were renamed. 3D nodes had a ``3D``
+suffix added to them for consistency with their 2D counterparts. This conversion
+is automatic.
+
+For ease of searching, this table lists all nodes and resources that were
+renamed in a way that does not involve only adding a ``3D`` suffix to the old name:
```
Comment (akien-mga) [inline tutorials/editor/upgrading_to_godot_4.rst:104]:
```suggestion
The list below refers to nodes which were simply renamed for consistency or
clarity in Godot 4.0. The project upgrade tool renames them automatically in
your scripts.

One noteworthy set of renames is 3D nodes, which all got a ``3D`` suffix added for
consistency with their 2D counterparts. For example, ``Area`` is now ``Area3D``.

For ease of searching, this table lists all nodes and resources that were renamed
and are automatically converted, excluding the ones which only involved adding
a ``3D`` suffix to the old name:
```

File: tutorials/editor/upgrading_to_godot_4.rst [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
++-----------------------------------------+-------------------------------------------+
+| VisualShaderNodeVec3Uniform             | VisualShaderNodeVec3Parameter             |
++-----------------------------------------+-------------------------------------------+
+| VisualShaderNodeUniform                 | VisualShaderNodeParameter                 |
++-----------------------------------------+-------------------------------------------+
+| VisualShaderNodeUniformRef              | VisualShaderNodeParameterRef              |
++-----------------------------------------+-------------------------------------------+
+
+If you cannot find a node or resource in the list below, refer to the above
+table to find its new name.
```
Comment (akien-mga) [inline tutorials/editor/upgrading_to_godot_4.rst:267]: The meaning of this sentence isn't really clear IMO. As a reader I don't know yet what "the list below" is, so it's confusing to be told to check the table above to find things if I can't find them...

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
Looks good to me, I'll let it stay open another day in case someone notices something else and will merge tomorrow. Thanks! ❤️

File: tutorials/editor/upgrading_to_godot_4.rst (hunk 8 above)
Comment (akien-mga) [reply tutorials/editor/upgrading_to_godot_4.rst:104]: I tried to make the language clearer that this is *not* all renamed nodes. IMO it's still a bit confusing though to have an "exhaustive" list that actually is only a subset.

There would need to be an intro paragraph to explain how the document is structured further (trivial renames handled by the converter, API changes which require manual handling, etc.).

But maybe there should be a table which actually lists *everything*, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

**Review by akien-mga — COMMENTED** [review]
Didn't review everything, but here's some early comments.

File: tutorials/editor/upgrading_to_godot_4.rst (hunk 7 above)
Comment (Calinou) [reply tutorials/editor/upgrading_to_godot_4.rst:6]: That was a leftover I didn't intend to keep :slightly_smiling_face: Fixed.

File: tutorials/editor/upgrading_to_godot_4.rst (hunk 8 above)
Comment (Calinou) [reply tutorials/editor/upgrading_to_godot_4.rst:104]: The table would be very long and could get outdated quickly – I guess we could find a way to generate it automatically and put it on a separate page.

Comment (Mickeon) [reply tutorials/editor/upgrading_to_godot_4.rst:104]: > But maybe there should be a table which actually lists everything, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

Absolutely. 100%. It's part of the reason as to why I expressed this:
> I was considering reordering that whole list 'cause it very much is... such a mess to read,

**Comment by Calinou** [comment]
> Some miscellaneous compatibility-breaking PRs that I could actually understand the meaning of, that I found while browsing.

I updated the list to take the PRs you listed into account.

**Review by akien-mga — APPROVED** [review]
Looks great as a start :+1:

**Comment by akien-mga** [comment]
Thanks!

**Comment by tinmanjuggernaut** [comment]
Good start. A comprehensive list of renames is great, but that's appendix information. The value of a guide like this is to discuss fundamentals, API design shifts, philosophical adjustments, and intentional breaks where there is no supported replacement or where reimplementation is required (eg tweens, dropping rotation_degrees), such as:
* Shaders are not converted at all with the auto tool (last I checked) and there are many breaking changes. Transmission is gone, and backlight works differently. Filenames are different. Hints are renamed. Projection Matrices are renamed. [#63673](https://github.com/godotengine/godot/issues/63673)
* ArrayMesh is an unsupported format, no matter if it is saved to disk as a resource file or embedded in a scene file. It is not forwards compatible and not upgradeable. The conversion tool won't do anything to the meshes. The only supported, and safe way to maintain working meshes is to keep fbx/glb files linked to tscns. Cleared inheritance is not supported at all and will cause broken meshes. GD3->4 [#63550](https://github.com/godotengine/godot/issues/63550) GD4a1-14->a15+ [#64854](https://github.com/godotengine/godot/issues/64854) There has been talk of supporting ArrayMesh by Juan, Remi & Clay in 64854, but until it is reality, the above should be documented.
* Advantages of upgrading: 
   * Vulkan support
   * Greatly improved visual fidelity with improved direct lighting & shadows, bounce lighting systems (VoxelGI/SDGFI), fixed roughness, improved SSS, physical lights and camera, volumetric fog, particle collision, vfx/screen space/camera effects, AA methods [Ground truth #63374](https://github.com/godotengine/godot/issues/63374)
   * Faster real world gdscript and rendering performance, even pre-optimizations (though not in all test circumstances [#68959](https://github.com/godotengine/godot/issues/68959))
   * Compute shaders and a much more open shader pipeline (How open is it? Can we do anything like tessellation or geometry shaders?), shader includes, uniform arrays, structs
   * Automatic, toggleable LOD generation and handling for meshes
   * Animation Library, allowing seamless retargeting of animations on identical rigs
   * Cyclic References of class names are now supported (as of pending beta7) [#67714](https://github.com/godotengine/godot/pull/67714)
   * Double precision floating point support (upon recompiling) [#892](https://github.com/godotengine/godot-proposals/issues/892#issuecomment-1046393101)
   * Move maker mode for offline, performance-independent rendering [#62122](https://github.com/godotengine/godot/pull/62122)
* DIsadvantages of upgrading:
   * Bullet physics has been removed. Only godot physics exists (for better or worse). (Gamedevs should not have to discover this half-way through a port.)
   * Any ArrayMeshes (cleared inheritance from glb/fbxes) will need to be manually reimported from assets again.
   * Rotation_degrees is gone affecting both code and AnimationPlayer tracks [#63390](https://github.com/godotengine/godot/issues/63390)
   * You now have to care about StringNames [#64171](https://github.com/godotengine/godot/issues/64171) Pending removal [#68747](https://github.com/godotengine/godot/pull/68747)
   * No 16/32-bit/HDR viewports (yet) [#2935](https://github.com/godotengine/godot-proposals/issues/2935)
   * Tween Node is gone and tweens need to be reimplemented, which can be done while still on GD3 [#41794](https://github.com/godotengine/godot/pull/41794)

**Pull request merged**